src/substrait/common/Io.cpp

```diff
@@ -0,0 +1,77 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+
+#include "substrait/common/Io.h"
+
+#include <regex>
+#include <string_view>
+
+#include "substrait/proto/plan.pb.h"
+#include "substrait/textplan/converter/LoadBinary.h"
+#include "substrait/textplan/converter/SaveBinary.h"
+#include "substrait/textplan/parser/LoadText.h"
+
+namespace io::substrait {
+
+namespace {
+
+const std::regex kIsJson(R"(("extensionUris"|"extensions"|"relations"))");
+const std::regex kIsProtoText(
+    R"((^|\n)((relations|extensions|extension_uris|expected_type_urls) \{))");
+const std::regex kIsText(
+    R"((^|\n) *(pipelines|[a-z]+ *relation|schema|source|extension_space) *)");
+
+PlanFileEncoding detectEncoding(std::string_view content) {
+  if (std::regex_search(content.begin(), content.end(), kIsJson)) {
+    return kJson;
+  }
+  if (std::regex_search(content.begin(), content.end(), kIsProtoText)) {
+    return kProtoText;
+  }
+  if (std::regex_search(content.begin(), content.end(), kIsText)) {
+    return kText;
+  }
+  return kBinary;
+}
+
+} // namespace
+
+absl::StatusOr<::substrait::proto::Plan> loadPlanWithUnknownEncoding(
+    std::string_view input_filename) {
+  auto contentOrError = textplan::readFromFile(input_filename.data());
```

Comment:
This does force a plan to be loaded fully into memory.  Is streaming reads something we could add support for in the future should it be needed?  At least, we should add this to a comment somewhere.

Reply:
Documented.